src/daemon/tabswitcher.vala

```diff
@@ -149,35 +188,120 @@ namespace Budgie {
+			if (window.is_skip_pager() || window.is_skip_tasklist()) return;
+
+			var window_widget = new TabSwitcherWidget(app_system, window);
+			var id = window_widget.id;
+
+			ids.insert(id, window_widget);
+
+			if (get_position_in_recency(id) == -1) recency.append(id);
+			window_box.insert(window_widget, -1);
+
+			window_box.set_max_children_per_line(ids.size() < 8 ? ids.size() : 8);
+
+			window_widget.window_activated.connect(set_window_as_activated);
+
+			window_widget.closed.connect(remove_window);
+
+			window_widget.workspace_changed.connect(() => {
+				window_box.invalidate_filter(); // Re-filter, maybe window is now on active workspace
+			});
+
+			window_box.invalidate_filter(); // Re-filter
+			window_box.invalidate_sort(); // Re-sort
+
+			queue_resize();
+		}
+
+		private bool flowbox_filter(FlowBoxChild box_child) {
+			TabSwitcherWidget? tab =  box_child.get_child() as TabSwitcherWidget;;
+
+			if ((tab == null) || (tab.window == null)) return false; // Hide if we can't cast
+
+			if (tab.window.is_skip_pager() || tab.window.is_skip_tasklist()) return false;
+			if (show_all_windows) return true;
+
+			return window_on_active_workspace(tab.window);
+		}
+
+		private int flowbox_sort(FlowBoxChild child1, FlowBoxChild child2) {
+			TabSwitcherWidget? tab1 =  child1.get_child() as TabSwitcherWidget;
+			TabSwitcherWidget? tab2 =  child2.get_child() as TabSwitcherWidget;
+			int64 pos1 = get_position_in_recency(tab1.id);
+			int64 pos2 = get_position_in_recency(tab2.id);
+			return pos1 < pos2 ? -1 : 1;
+		}
+
+		// get_position_in_recency will get the position in recency
+		// You might think "hey, but List has an index function, just use that!"
+		// No, because for whatever reason it fails at such a basic task
+		private int64 get_position_in_recency(string id) {
+			for (int i = 0; i < recency.length(); i++) {
+				var val = recency.nth_data((uint) i);
+				if (val == id) return i;
+			}
+
+			return -1;
+		}
+
 		/**
-		* Monitors changed, find out the primary monitor, and schedule move of OSD
+		* Make the current selection the active window
 		*/
+		private void on_hide() {
+			var selection = window_box.get_selected_children();
+			Gtk.FlowBoxChild? current = null;
+			if (selection != null && !selection.is_empty()) {
+				current = selection.nth_data(0) as Gtk.FlowBoxChild;
+			}
+
+			if (current == null) return;
+
+			/* Get the window, which should be activated and activate that */
+			TabSwitcherWidget? tab = current.get_child() as TabSwitcherWidget;
+			window_box.unselect_child(current);
+
+			try {
+				libxfce4windowing.Workspace? workspace = tab.window.get_workspace();
+				if (workspace != null) workspace.activate();
+			} catch (GLib.Error e) {
+				stderr.printf("Failed to activate workspace. Window switching will almost certainly fail: %s\n", e.message);
+			}
+
+			try {
+				tab.window.activate(get_time());
+			} catch (GLib.Error e) {
+				stderr.printf("Failed to activate window: %s\n", e.message);
```

Comment:
Can we not use the GLib logging functions here? I know places in daemon can't for some reason, just making sure this is one of them.

Reply:
Changing it here is probably fine but will check. Sadly it's very easy for me to trigger activation failure (looking at you Visual Studio Code).

Reply:
Should be sorted as of cdb3a5e